src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs

```diff
@@ -196,13 +249,26 @@ private async Task ProcessQueueAsync()
                         // will be sent back to the client but they can also be captured via HandleNonMutatingRequestError,
                         // though these errors don't put us into a bad state as far as the rest of the queue goes.
                         // Furthermore we use Task.Run here to protect ourselves against synchronous execution of work
-                        // blocking the request queue for longer periods of time (it enforces parallelizabilty).
-                        _ = WrapStartRequestTaskAsync(Task.Run(() => work.StartRequestAsync(context, cancellationToken), cancellationToken), rethrowExceptions: false);
+                        // blocking the request queue for longer periods of time (it enforces parallelizability).
+                        var pendingTask = WrapStartRequestTaskAsync(Task.Run(() => work.StartRequestAsync(context, cancellationToken), cancellationToken), rethrowExceptions: false);
+
+                        if (CancelInProgressWorkUponMutatingRequest)
+                        {
+                            pendingTasks.TryAdd(pendingTask, cancellationTokenSource);
+
+                            _ = pendingTask.ContinueWith(t =>
+                            {
+                                if (pendingTasks.TryRemove(t, out var pendingCancellationTokenSource))
```

Comment:
feels like this shuold just be Contract.ThrowIfFalse.  this legit acn't fail (unless something horrific happened).

Reply:
I don't think the Roslyn.Utility.Contract class is available from this code. I added explicit InvalidOperationExceptions for these conditions.